File: app/src/main/java/com/defianttech/convertme/UnitCollection.java
package com.defianttech.convertme;

import android.content.Context;
import androidx.annotation.NonNull;

import android.content.SharedPreferences;
import android.util.Log;

import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

/*
 * Copyright (c) 2014-2017 Dmitry Brant
 */
public class UnitCollection {
    private static final String TAG = "UnitCollection";
    private static final String CUSTOM_COLLECTION_PREF_NAME = "custom_collection";

    public static final int DEFAULT_CATEGORY = 5; //default to "distance"
    public static final int DEFAULT_FROM_INDEX = 10; //default to "inch"
    public static final int DEFAULT_TO_INDEX = 2; //default to "centimeter"
    public static final double DEFAULT_VALUE = 1.0;
    public static final int CUSTOM_ID_START = 10000;

    private static UnitCollection[] INSTANCE;
    private static String[] allCategoryNames;

    @NonNull
    public static UnitCollection[] getInstance(@NonNull Context context) {
        if (INSTANCE == null) {
            resetInstance(context);
        }
        return INSTANCE;
    }

    private static void resetInstance(@NonNull Context context) {
        INSTANCE = getAllUnits(context);
    }

    @NonNull
    public static String[] getAllCategoryNames(@NonNull Context context) {
        if (allCategoryNames == null) {
            UnitCollection[] collections = getInstance(context);
            List<String> unitCategories = new ArrayList<>();
            for (UnitCollection collection : collections) {
                unitCategories.addAll(Arrays.asList(collection.getNames()));
            }
            allCategoryNames = unitCategories.toArray(new String[unitCategories.size()]);
            Arrays.sort(allCategoryNames, new Comparator<String>() {
                @Override
                public int compare(String left, String right) {
                    return left.compareTo(right);
                }
            });
        }
        return allCategoryNames;
    }

    private final String[] names;
    public String[] getNames() {
        return names;
    }

    private final List<SingleUnit> items;
    public List<SingleUnit> getItems() {
        return items;
    }

    public SingleUnit get(int index) {
        return items.get(index);
    }

    public int length() {
        return items.size();
    }

    public static int collectionIndexByName(UnitCollection[] collections, String name) {
        for (int i = 0; i < collections.length; i++) {
            for (String cName : collections[i].getNames()) {
                if (cName.equals(name)) {
                    return i;
                }
            }
        }
        return 0;
    }

    public static double convert(@NonNull Context context, int category, int fromIndex, int toIndex, double value) {
        UnitCollection[] collections = getInstance(context);
        double result = (value - collections[category].get(fromIndex).getOffset())
                / collections[category].get(fromIndex).getMultiplier();
        result *= collections[category].get(toIndex).getMultiplier();
        result += collections[category].get(toIndex).getOffset();
        return result;
    }

    UnitCollection(String[] collectionNames, List<SingleUnit> collectionItems) {
        names = collectionNames;
        items = collectionItems;
    }

    @NonNull
    static UnitCollection[] getAllUnits(@NonNull Context context) {
        List<UnitCollection> collections = new ArrayList<>();
        InputStream inStream = null;
        try {
            Log.d(TAG, "Loading units from assets...");
            inStream = context.getAssets().open("units.txt");
            BufferedReader reader = new BufferedReader(new InputStreamReader(inStream));
            List<SingleUnit> currentCollection = new ArrayList<>();
            String line;
            String[] lineArr;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.startsWith("#") || line.length() == 0) {
                    continue;
                }
                if (line.startsWith("==")) {
                    currentCollection = new ArrayList<>();
                    lineArr = line.replace("==", "").trim().split("\\s*,\\s*");
                    collections.add(new UnitCollection(lineArr, currentCollection));
                    continue;
                }
                lineArr = line.split("\\s*,\\s*");
                currentCollection.add(new SingleUnit(Integer.parseInt(lineArr[0]), lineArr[1], Double.parseDouble(lineArr[2]), Double.parseDouble(lineArr[3])));
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to read unit collection.", e);
        } finally {
            if (inStream != null) {
                try { inStream.close(); }
                catch(Exception e) {
                    //
                }
            }
        }

        // deserialize and append custom units.
        CustomUnits customUnits = getCustomUnits(context);

        for (CustomUnits.CustomUnit unit : customUnits.getUnits()) {
            UnitCollection category = collections.get(unit.getCategoryId());
            SingleUnit baseUnit = null;
            for (SingleUnit u : category.getItems()) {
                if (u.getId() == unit.getBaseUnitId()) {
                    baseUnit = u;
                    break;
                }
            }
            if (baseUnit == null) {
                break;
            }
            category.getItems().add(new SingleUnit(unit.getId(), unit.getName(),
                    baseUnit.getMultiplier() * unit.getMultiplier(), unit.getOffset()));
        }

        return collections.toArray(new UnitCollection[0]);
    }

    @NonNull
    static CustomUnits getCustomUnits(@NonNull Context context) {
        SharedPreferences prefs = ConvertActivity.getPrefs(context);
        String customSerialized = prefs.getString(CUSTOM_COLLECTION_PREF_NAME, "{}");
        CustomUnits customUnits = new Gson().fromJson(customSerialized, CustomUnits.class);
        return customUnits != null ? customUnits : new CustomUnits();
    }

    static void saveCustomUnits(@NonNull Context context, CustomUnits customUnits) {
        SharedPreferences.Editor editor = ConvertActivity.getPrefs(context).edit();
        editor.putString(CUSTOM_COLLECTION_PREF_NAME, new Gson().toJson(customUnits));
        editor.apply();
    }

    static void addCustomUnit(@NonNull Context context, int categoryId, int baseUnitId, double multiplier, String name) {
        CustomUnits customUnits = getCustomUnits(context);
        int maxId = CUSTOM_ID_START;
        for (CustomUnits.CustomUnit u : customUnits.getUnits()) {
            if (u.getId() >= maxId) {
                maxId = u.getId() + 1;
            }
        }
        CustomUnits.CustomUnit unit = new CustomUnits.CustomUnit(maxId, categoryId, baseUnitId, 0.0, multiplier, name);
        List<CustomUnits.CustomUnit> newUnits = new ArrayList<>(customUnits.getUnits());
        newUnits.add(unit);
        customUnits.setUnits(newUnits);
        saveCustomUnits(context, customUnits);
        resetInstance(context);
    }

    static void deleteCustomUnit(@NonNull Context context, CustomUnits.CustomUnit unit) {
        CustomUnits customUnits = getCustomUnits(context);
        List<CustomUnits.CustomUnit> newUnits = new ArrayList<>(customUnits.getUnits());
        int index = -1;
        for (int i = 0; i < newUnits.size(); i++) {
            if (newUnits.get(i).getId() == unit.getId()) {
                index = i;
                break;
            }
        }
        if (index >= 0) {
            newUnits.remove(index);
        }
        customUnits.setUnits(newUnits);
        saveCustomUnits(context, customUnits);
        resetInstance(context);
    }
}

File: app/src/main/java/com/defianttech/convertme/UnitCollection.kt
package com.defianttech.convertme

import android.content.Context
import android.util.Log
import com.defianttech.convertme.CustomUnits.CustomUnit
import com.google.gson.Gson
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import kotlin.math.max

/*
* Copyright (c) 2014-2022 Dmitry Brant
*/
class UnitCollection internal constructor(val names: List<String>, val items: MutableList<SingleUnit>) {
    operator fun get(index: Int): SingleUnit {
        return items[index]
    }

    fun length(): Int {
        return items.size
    }

    companion object {
        private const val TAG = "UnitCollection"
        private const val CUSTOM_COLLECTION_PREF_NAME = "custom_collection"
        private val sRegex = "\\s*,\\s*".toRegex()

        const val DEFAULT_CATEGORY = 5 //default to "distance"
        const val DEFAULT_FROM_INDEX = 10 //default to "inch"
        const val DEFAULT_TO_INDEX = 2 //default to "centimeter"
        const val DEFAULT_VALUE = 1.0

        private const val CUSTOM_ID_START = 10000

        private var INSTANCE: List<UnitCollection>? = null
        private var ALL_CATEGORY_NAMES: List<String>? = null

        fun getInstance(context: Context): List<UnitCollection> {
            if (INSTANCE == null) {
                resetInstance(context)
            }
            return INSTANCE!!
        }

        private fun resetInstance(context: Context) {
            INSTANCE = getAllUnits(context)
        }

        fun getAllCategoryNames(context: Context): List<String> {
            if (ALL_CATEGORY_NAMES == null) {
                ALL_CATEGORY_NAMES = getInstance(context).flatMap { it.names }.sortedBy { it }
            }
            return ALL_CATEGORY_NAMES!!
        }

        fun collectionIndexByName(collections: List<UnitCollection>, name: String): Int {
            for (i in collections.indices) {
                if (collections[i].names.find { it == name } != null) {
                    return i
                }
            }
            return 0
        }

        fun convert(context: Context, category: Int, fromIndex: Int, toIndex: Int, value: Double): Double {
            val collections = getInstance(context)
            var result = ((value - collections[category][fromIndex].offset)
                    / collections[category][fromIndex].multiplier)
            result *= collections[category][toIndex].multiplier
            result += collections[category][toIndex].offset
            return result
        }

        private fun getAllUnits(context: Context): List<UnitCollection> {
            val collections = mutableListOf<UnitCollection>()
            var inStream: InputStream? = null
            try {
                Log.d(TAG, "Loading units from assets...")
                inStream = context.assets.open("units.txt")
                val reader = BufferedReader(InputStreamReader(inStream))
                var currentCollection = mutableListOf<SingleUnit>()
                var line: String
                var lineArr: List<String>
                while (reader.readLine().also { line = it.orEmpty() } != null) {
                    line = line.trim()
                    if (line.startsWith("#") || line.isEmpty()) {
                        continue
                    }
                    if (line.startsWith("==")) {
                        currentCollection = mutableListOf()
                        lineArr = line.replace("==", "").trim().split(sRegex)
                        collections.add(UnitCollection(lineArr, currentCollection))
                        continue
                    }
                    lineArr = line.split(sRegex)
                    currentCollection.add(SingleUnit(lineArr[0].toInt(), lineArr[1], lineArr[2].toDouble(), lineArr[3].toDouble()))
                }
            } catch (e: IOException) {
                Log.e(TAG, "Failed to read unit collection.", e)
            } finally {
                if (inStream != null) {
                    try {
                        inStream.close()
                    } catch (e: Exception) {
                        //
                    }
                }
            }

            // deserialize and append custom units.
            val customUnits = getCustomUnits(context)
            for (unit in customUnits.units) {
                val category = collections[unit.categoryId]
                val baseUnit = category.items.find { it.id == unit.baseUnitId } ?: break
                category.items.add(SingleUnit(unit.id, unit.name,
                        baseUnit.multiplier * unit.multiplier, unit.offset))
            }
            return collections
        }

        fun getCustomUnits(context: Context): CustomUnits {
            val prefs = ConvertActivity.getPrefs(context)
            val customSerialized = prefs.getString(CUSTOM_COLLECTION_PREF_NAME, "{}")
            var customUnits: CustomUnits? = null
            try {
                customUnits = Gson().fromJson(customSerialized, CustomUnits::class.java)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return customUnits ?: CustomUnits()
        }

        private fun saveCustomUnits(context: Context, customUnits: CustomUnits?) {
            val editor = ConvertActivity.getPrefs(context).edit()
            editor.putString(CUSTOM_COLLECTION_PREF_NAME, Gson().toJson(customUnits))
            editor.apply()
        }

        fun addCustomUnit(context: Context, categoryId: Int, baseUnitId: Int, multiplier: Double, offset: Double, name: String) {
            val customUnits = getCustomUnits(context)
            val maxId = max(CUSTOM_ID_START,(customUnits.units.maxOfOrNull { it.id } ?: 0) + 1)
            val unit = CustomUnit(maxId, categoryId, baseUnitId, offset, multiplier, name)
            customUnits.units.add(unit)
            saveCustomUnits(context, customUnits)
            resetInstance(context)
        }

        fun editCustomUnit(context: Context, id: Int, baseUnitId: Int, multiplier: Double, name: String) {
            val customUnits = getCustomUnits(context)
            val unit = customUnits.units.find { it.id == id } ?: return
            unit.name = name
            unit.baseUnitId = baseUnitId
            unit.multiplier = multiplier
            saveCustomUnits(context, customUnits)
            resetInstance(context)
        }

        fun deleteCustomUnit(context: Context, unit: CustomUnit) {
            val customUnits = getCustomUnits(context)
            var index = -1
            for (i in customUnits.units.indices) {
                if (customUnits.units[i].id == unit.id) {
                    index = i
                    break
                }
            }
            if (index >= 0) {
                customUnits.units.removeAt(index)
            }
            saveCustomUnits(context, customUnits)
            resetInstance(context)
        }
    }
}
